app/src/swap.c

```diff
@@ -0,0 +1,290 @@
+/*******************************************************************************
+*   (c) 2016 Ledger
+*   (c) 2018 - 2023 Zondax AG
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+********************************************************************************/
+#include "swap.h"
+#include "crypto.h"
+#include "bignum.h"
+#include "zxformat.h"
+#include "substrate_dispatch.h"
+
+static zxerr_t extractHDPath(uint8_t* params, uint8_t paramsSize);
+static zxerr_t readU32BE(uint8_t* input, uint32_t *output);
+static zxerr_t readU64BE(uint8_t* input, uint64_t *output);
+static zxerr_t bytesAmountToStringBalance(uint8_t *amount, uint8_t amount_len, char *out, uint8_t out_len);
+
+static void handle_check_address(check_address_parameters_t* params);
+static void handle_get_printable_amount(get_printable_amount_parameters_t* params);
+
+static bool copy_transaction_parameters(create_transaction_parameters_t* sign_transaction_params);
+
+swap_globals_t G_swap_state;
+
+void swap_handle_command(libargs_s *args) {
+    switch (args->command) {
+        case check_address:
+            handle_check_address(args->check_address);
+            G_swap_state.called_from_swap = 0;
+            break;
+
+        case sign_transaction:
+            if (copy_transaction_parameters(args->create_transaction)) {
```

Comment:
what's the expected behaviour if copy_transaction_parameters returns false? Run app or exiting with an error?

Reply:
@agrojean-ledger

Reply:
@ftheirs @carlosala If `copy_transaction_parameters` `os_lib_end` is called and the execution returns to the app-exchange. After that I'm not sure how it will fail (gracefully or not) but it will surely fail because the transaction APDU won't be processed by Polkadot... In other apps with the swap, it's done in the same way.

It's a good question though, how do we handle that on the app-exchange side ? But out of the scope of this PR IMO.

Reply:
@agrojean-ledger then we should set called from swap to 0 in order to make `os_lib_end` be executed [here](https://github.com/Zondax/ledger-polkadot/blob/5b2d4d1bf671c9e1861238b73a2e5faaa5c15bbb/app/src/common/main.c#L51), right?

Reply:
@carlosala I've just been thinking about that we also need to make sure the app exits properly after signing the transaction APDU during a swap. I'm not sure we've done that.

Reply:
@agrojean-ledger yeah, it seems important. Could you review that and suggest how to proceed?

Reply:
We should set **G_swap_state.should_exit = 1** after signing the transaction and that will be handled in **FINALLY** block in app_main. If called_from_swap and should_exit, then os_lib_end() will be called

Reply:
@ftheirs @carlosala in some apps (i.e. Ethereum) , the app is exited as soon as the `io_exchange` call to send the signature is done : 

`   
    // ...
    // ... Do stuff (prepare signature and fill io buffer... 
    // ...
    io_exchange(CHANNEL_APDU | IO_RETURN_AFTER_TX, tx);
    if (called_from_swap) {
        os_sched_exit(0);
    }
`

Reply:
One way to do it simply would be to call `os_sched_exit(0)` in  `handleSignEd25519` and `handleSignSr25519` but maybe you won't find it clean enough : 

```__Z_INLINE void handleSignEd25519(volatile uint32_t *flags, volatile uint32_t *tx) {
    
   // ...
   // ... Do stuff ...
   // ...

    if (G_swap_state.called_from_swap) {
        app_sign_ed25519();
        // Exit app, our job is done.
         os_sched_exit(0);
    } else {
        view_review_init(tx_getItem, tx_getNumItems, app_sign_ed25519);
        view_review_show(REVIEW_TXN);
        *flags |= IO_ASYNCH_REPLY;
    }
}
```

Reply:
@agrojean-ledger this has been done 👌🏻